balancer/weightedroundrobin/balancer.go

```diff
@@ -141,273 +147,311 @@ func (bb) Name() string {
+			ew := ewv.(*endpointWeight)
+			readyPickersWeight = append(readyPickersWeight, pickerWeightedEndpoint{
+				picker:           childState.State.Picker,
+				weightedEndpoint: ew,
+			})
 		}
-		// Update config for existing weightedSubConn or send update for first
-		// time to new one.  Ensures an OOB listener is running if needed
-		// (and stops the existing one if applicable).
-		wsc.updateConfig(b.cfg)
+	}
+	// If no ready pickers are present, simply defer to the round robin picker
+	// from endpoint sharding, which will round robin across the most relevant
+	// pick first children in the highest precedence connectivity state.
+	if len(readyPickersWeight) == 0 {
+		b.ClientConn.UpdateState(balancer.State{
+			ConnectivityState: state.ConnectivityState,
+			Picker:            state.Picker,
+		})
+		return
 	}
 
-	// Loop through existing subconns and remove ones that are not in addrs.
-	for _, addr := range b.subConns.Keys() {
-		if _, ok := addrsSet.Get(addr); ok {
-			// Existing address also in new address list; skip.
-			continue
-		}
-		// addr was removed by resolver.  Remove.
-		wsci, _ := b.subConns.Get(addr)
-		wsc := wsci.(*weightedSubConn)
-		wsc.SubConn.Shutdown()
-		b.subConns.Delete(addr)
+	p := &picker{
+		v:               rand.Uint32(), // start the scheduler at a random point
+		cfg:             b.cfg,
+		weightedPickers: readyPickersWeight,
+		metricsRecorder: b.metricsRecorder,
+		locality:        b.locality,
+		target:          b.target,
 	}
+
+	var ctx context.Context
+	ctx, b.stopPicker = context.WithCancel(context.Background())
+	p.start(ctx)
+
+	b.ClientConn.UpdateState(balancer.State{
+		ConnectivityState: state.ConnectivityState,
+		Picker:            p,
+	})
 }
 
-func (b *wrrBalancer) ResolverError(err error) {
-	b.resolverErr = err
-	if b.subConns.Len() == 0 {
-		b.connectivityState = connectivity.TransientFailure
+type pickerWeightedEndpoint struct {
+	picker           balancer.Picker
+	weightedEndpoint *endpointWeight
+}
+
+func (b *wrrBalancer) NewSubConn(addrs []resolver.Address, opts balancer.NewSubConnOptions) (balancer.SubConn, error) {
+	addr := addrs[0] // The new pick first policy for DualStack will only ever create a SubConn with one address.
+	var sc balancer.SubConn
+
+	oldListener := opts.StateListener
+	opts.StateListener = func(state balancer.SubConnState) {
+		b.updateSubConnState(sc, state)
+		oldListener(state)
 	}
-	if b.connectivityState != connectivity.TransientFailure {
-		// No need to update the picker since no error is being returned.
-		return
+
+	sc, err := b.ClientConn.NewSubConn([]resolver.Address{addr}, opts)
+	if err != nil {
+		return sc, err
+	}
+
+	b.mu.Lock()
+	ewv, ok := b.addressWeights.Get(addr)
+	if !ok {
+		b.mu.Unlock()
+		// SubConn state updates can come in for a no longer relevant endpoint
+		// weight (from the old system after a new config update is applied).
+		// Will eventually get cleared from scMap once receives Shutdown signal.
+		return sc, err
```

Comment:
As above, but moreso..return `nil, err` -- your `sc` is actually valid here.

Reply:
Ah ok. I deleted the last part of comment and moved it to updateSubConnState since that actually applies to updateSubConnState.

Reply:
Don't I still want to return the created SubConn to the child though? Even if it's from an old system/irrelevant. I guess it doesn't matter since that child will eventually be cleared anyway and this layer never tracks it.

Reply:
So I induced a nil panic from this change to return nil and I realized why, this caught a bug.

So in updateEndpoints say I got a update with endpoint with address 1. Then I got an endpoint with address 1 and 2. In the cleanup step, it would cleanup the old endpoints addresses even if present in the new one, leading to hitting this block and inducing a nil panic. Fixing that by tracking a current address map fixed it.

Reply:
Wait, isn't this bad?  We need to do the lookup before calling `NewSubConn` or else you're creating a subchannel you don't actually need.  In which cases it would need to be `Shutdown()` first.

Reply:
Don't we want this to still be created, and eventually shutdown from the old child balancer going away? Why gate here? Isn't this an implicit part of the child balancers lifecycle? (When it goes away and Close() since for an irrelevant endpoint), it sends a Shutdown signal for it's created SC's? How will this process occur if we gate this creation here?